_posts/2019-12-23-Presto-Experiment-with-Graivton-Processor.md

```diff
@@ -0,0 +1,162 @@
+
+But is it true? Initially, we do not have a clear answer to how much effort we need to bring Presto into the world of the different processors. No care about the underlying platform is generally beneficial for development. But if using different processors enables us to accelerate the performance and stability of Presto, we must care about it. We must prove anything unclear by the experiment.
+
+This article is the report to clarify what we need to do to run Presto on the Arm-based platform and see how much benefit we can potentially obtain with Graviton Processor.
+
+As the Graviton 2 based instance types are preview state, We tried to run Presto on A1 instance that has the first generation of Graviton processor inside. It still would be a helpful anchor to understand the potential benefit of the Graviton 2 processor.
+
+<!--more-->
+
+# How to make Presto compatible with Arm
+
+We are going to build the binary of Presto supporting Arm platform first. From the results, there are not so many things to do so. As long as JVM supports the Arm platform, it should work without any modification in the application code. But Presto has some restrictions on the platform where it runs to protect the functionality, including plugins. For instance, the latest Presto supports only [x86 and PowerPC architectures](https://github.com/prestosql/presto/blob/ee05ee5221690d66598039c6e397f7c7cb4c202b/presto-main/src/main/java/io/prestosql/server/PrestoSystemRequirements.java#L69). This limitation prevents us from using Presto on the Arm platform.
+
+To make Presto runnable on Arm machine, we need to modify [PrestoSystemRequirements](https://github.com/prestosql/presto/blob/master/presto-main/src/main/java/io/prestosql/server/PrestoSystemRequirements.java) class to allow `aarch64` architecture and more. For experimental purposes, we can apply such a patch to remove the restriction altogether.
+
+```
+diff --git a/presto-main/src/main/java/io/prestosql/server/PrestoSystemRequirements.java b/presto-main/src/main/java/io/prestosql/server/PrestoSystemRequirements.java
+index 07b7d12c64..b6a1249681 100644
+--- a/presto-main/src/main/java/io/prestosql/server/PrestoSystemRequirements.java
++++ b/presto-main/src/main/java/io/prestosql/server/PrestoSystemRequirements.java
+@@ -71,9 +71,9 @@ final class PrestoSystemRequirements
+ String osName = StandardSystemProperty.OS_NAME.value();
+ String osArch = StandardSystemProperty.OS_ARCH.value();
+ if ("Linux".equals(osName)) {
+- if (!"amd64".equals(osArch) && !"ppc64le".equals(osArch)) {
+- failRequirement("Presto requires amd64 or ppc64le on Linux (found %s)", osArch);
+- }
+ if ("ppc64le".equals(osArch)) {
+ warnRequirement("Support for the POWER architecture is experimental");
+ }
+```
+
+This patch is all we have to do to run Presto on the Arm platform. It should work for most cases.
```

Comment:
We should mention that the Hive connector does not work, as it has native code not yet available for that platform.